### PR TITLE
Added best practices for application and operator development

### PR DESCRIPTION
**Pull request by pramodin**
This was originally submitted to Malhar, pull request mentioned below. This is a more appropriate repository because operator and application documentation is here so submitting it here.

https://github.com/apache/apex-malhar/pull/350

### CODE REVIEW

**Comment by pramodin** [comment]
@tweise @amberarrow please see

File: docs/development_best_practices.md [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+	* An optional concrete operator also in the library that extends the abstract operator and provides commonly used schema types such as strings, byte[] or POJOs.
+* Follow the conventions of the life cycle methods
+	* Do one time initialization of entities that apply for the entire lifetime of the operator in the **setup** method, e.g., factory initializations. Initializations in **setup** are done in the container where the operator is deployed. Initializations in the constructor are discouraged as they lead to extra garbage in memory. This reason is as follows: The operator is instantiated on the client from where the application is launched, then serialized and made available for launch. So the constructor is first called on the client and if it were to initialize any of the fields, that state would be saved during serialization. In the launch process the operator is deserialized. This would invoke the constructor again, which will initialize the fields but their state will get overwritten by the serialized state and the initial values would become garbage in memory.
```
Comment (amberarrow) [inline docs/development_best_practices.md]: You prefer "This reason ... " to "The reason ... " for what reason ? :-}

The proscription seems too broad: Initialization of primitive fields generate no garbage. Also, transient fields are not serialized. The reader, when she reads this, will naturally think about initialization at the point of definition; wouldn't a sentence or two addressing that case be worthwhile ?

File: docs/development_best_practices.md [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+	* Do one time initialization for live entities in **activate** method, e.g., opening connections to a database server or starting a thread for asynchronous operations. The **activate** method is called right before processing starts so it is a better place to do these kind of initializations than at **setup** which can lead to a delay before processing data from the live entity.
```
Comment (amberarrow) [inline docs/development_best_practices.md]: to do these kind of => for these

File: docs/development_best_practices.md [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+	* Do one time initialization for live entities in **activate** method, e.g., opening connections to a database server or starting a thread for asynchronous operations. The **activate** method is called right before processing starts so it is a better place to do these kind of initializations than at **setup** which can lead to a delay before processing data from the live entity.  
+	* Perform periodic tasks based on processing time in application window boundaries
```
Comment (amberarrow) [inline docs/development_best_practices.md]: Need terminating periods on many of these points.

**Comment by amberarrow** [comment]
Looks good. A very valuable addition to our docs -- I've seen developers run into these issues time and time again.

File: docs/development_best_practices.md (hunk 1 above)
Comment (pramodin) [reply docs/development_best_practices.md]: "The reason" trumps "this reason" and for that reason, I will change "this reason". I will reword it to indicate the specific initializations I am thinking about.

**Comment by tweise** [comment]
Looks good. Please rebase and squash. You can automatically close the other PR when this is merged by modifying the commit message: https://help.github.com/articles/closing-issues-via-commit-messages/

**Comment by pramodin** [comment]
@tweise done
